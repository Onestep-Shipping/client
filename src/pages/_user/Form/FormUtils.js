import CREATE_BOOKING_REQUEST from '../../../apollo/mutations/CreateBookingRequestMutation.js';
import CREATE_BILL_INSTRUCTION from '../../../apollo/mutations/CreateBillInstructionMutation.js';

const handleBookingRequest = (form, scheduleId, quoteId) => {
  const bookingRequest = (createBookingRequestFromForm(form));
  const variables = { 
    companyId: "5ebb4f56b6a43ab1f1500127",
    scheduleId, 
    quoteId,
    bookingRequest
  }
  return { 
    mutation: CREATE_BOOKING_REQUEST, variables 
  }
};

const createBookingRequestFromForm = form => {
  const { commodity, hsCode, quantity, containerType, paymentTerm, autoFilling } = form;
  const quantities = getValuesOfNodeList(quantity);
  const containerTypes = getValuesOfNodeList(containerType);

  const containers = [];
  for (let i = 0; i < quantities.length; i++) {
    containers.push({ 
      containerType: containerTypes[i], 
      quantity: parseInt(quantities[i]) 
    })
  }

  return {
    commodity: commodity.value,
    hsCode: hsCode.value,
    containers,
    paymentTerm: paymentTerm.value,
    autoFilling: autoFilling.value === "Yes",
  }
};

const handleBillInstruction = (form, shipmentId) => {
  const billInstruction = (createBillInstructionFromForm(form));
  const variables = {
    shipmentId,
    billInstruction
  }
  return { 
    mutation: CREATE_BILL_INSTRUCTION, variables 
  }
};

const createBillInstructionFromForm = form => {
  console.log(form);
  const { 
    shipper, consignee, notify, descriptionOfGoods,
    containerNo, seelNo, cargoWeight, measurement, vGM,
    orderPONumber, hSCode, cAEDAESNumber, cargoValue
  } = form;
  
  const containerNos = getValuesOfNodeList(containerNo);
  const seelNos = getValuesOfNodeList(seelNo);
  const cargoWeights = getValuesOfNodeList(cargoWeight);
  const measurements = getValuesOfNodeList(measurement);
  const vgms = getValuesOfNodeList(vGM);

  const containers = [];
  for (let i = 0; i < containerNos.length; i++) {
    containers.push({ 
      containerNo: containerNos[i],
      seelNo: seelNos[i],
      weight: parseFloat(cargoWeights[i]),
      measurement: parseFloat(measurements[i]),
      vgm: parseFloat(vgms[i]),
    })
  }

  return {
    shipper: shipper.value,
    consignee: consignee.value,
    notify: notify.value,
    description: descriptionOfGoods.value,
    containers,
    orderNo: orderPONumber.value,
    hsCode: hSCode.value,
    caedNo: cAEDAESNumber.value,
    cargoValue: cargoValue.value
  }
}

const getValuesOfNodeList = potentialList => {
  const result = [];
  if (potentialList.value === "") { // has a list of values
    potentialList.forEach(input => {
      if (input.value !== "") {
        result.push(input.value);
      }
    });
  } else { // only has one value
    result.push(potentialList.value);
  }
  return result;
}

export { handleBookingRequest, handleBillInstruction }