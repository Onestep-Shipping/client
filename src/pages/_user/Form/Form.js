import React, {useCallback} from 'react';
import './Form.css';
import { handleBookingRequest, handleBillInstruction } from './FormUtils.js';
import PropTypes from 'prop-types';

import {useParams, useHistory} from 'react-router-dom';
import Header from '../../../components/Header/Header.js';
import BookingDisplay from '../../../components/BookingDisplay/BookingDisplay.js';
import BookingForm from '../../../components/ServiceForm/BookingForm.js';
import BolForm from '../../../components/ServiceForm/BolForm.js';
import client from '../../../apollo/index.js';

const Form = (props) => {
  const { type, id } = useParams();
  const history = useHistory();
  const { schedule, quote } = props.location.state;

  const handleSubmit = useCallback((e) => {
    e.preventDefault();
    const { mutation, variables } = type === "booking" ? 
      handleBookingRequest(e.target, schedule._id, quote._id) : 
      handleBillInstruction(e.target, id); // shipment Id for Bill Instruction
    
    client.mutate({ mutation, variables })
      .then(res => {
        const { 
          createBookingRequestAndInitShipment, createBillInstruction 
        } = res.data;
        if (createBookingRequestAndInitShipment === "OK" || 
            createBillInstruction               === "OK") {
          history.push('/success/' + type + '/' + id);
        }  
      })
  }, [history, id, type]);

  const formBool = type === 'booking';
  const fieldNumber = formBool ? 8 : 3;

  return (
    <div className="homepage-container">
      <Header />
      <div className="body-container2">
        <div className="booking-container">
          <text className="schedule-header-text">
            {formBool ? 'Booking Request' : 'Bill Of Lading (BOL) Instruction'} 
          </text>
          <div className={"info-container" + (formBool ? "" : "2")}>
            <BookingDisplay schedule={schedule} quote={quote} fields={fieldNumber}/>
            {formBool ?
              <BookingForm action={handleSubmit} /> : 
              <BolForm action={handleSubmit} shipmentId={id} />
            }
          </div>
        </div>
      </div>
    </div>
  );
};

export default Form;

Form.propTypes = {
  formType: PropTypes.string,
  location: PropTypes.object,
};
