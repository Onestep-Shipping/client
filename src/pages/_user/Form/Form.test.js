import Form from './Form';
import { MemoryRouter } from 'react-router-dom';
import React from 'react';
import { shallow } from 'enzyme';

let wrapper = shallow(
  <MemoryRouter>
    <Form />
  </MemoryRouter>);

describe('Form', () => {
  it('should render the Form Component correctly', () => {   
    expect(Form).toMatchSnapshot();
  });
});