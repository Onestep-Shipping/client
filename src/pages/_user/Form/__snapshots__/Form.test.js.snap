// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render the Form Component correctly 1`] = `[Function]`;
